lib/puppet/rest_client/client.rb

```diff
@@ -0,0 +1,76 @@
+# frozen_string_literal: true
+require 'httpclient'
+require 'puppet'
+require 'puppet/file_system'
+
+module Puppet
+  module Rest
+    # TBD
+    class Client
+      def self.instance
+        @@instance ||= new(Puppet.settings[:http_user_agent],
```

Comment:
We should do better lifecycle management, a singleton doesn't seem appropriate. These client objects should live only for the duration of an agent run, and should use the same server and port for the whole lifetime.

Reply:
Because the certificate downloading happens outside the context of the actual agent run, the client responsible for those requests should probably share a lifetime with with the application context that triggers the bootstrapping.